src/rules/no-redundant-files.ts

```diff
@@ -0,0 +1,252 @@
+import type { Rule } from "eslint";
+import type { AST as JsonAST } from "jsonc-eslint-parser";
+import type { JSONStringLiteral } from "jsonc-eslint-parser/lib/parser/ast.js";
+
+import * as ESTree from "estree";
+
+import { createRule } from "../createRule.js";
+import { isJSONStringLiteral, isNotNullish } from "../utils/predicates.js";
+
+const defaultFiles = [
+	/* cspell:disable-next-line */
+	"LICENCE",
+	/* cspell:disable-next-line */
+	"LICENCE.md",
+	"LICENSE",
+	"LICENSE.md",
+	"package.json",
+	"README.md",
+] as const;
+
+const cachedRegex = new Map<string, RegExp>();
+const getCachedLocalFileRegex = (filename: string) => {
+	// Strip the leading `./`, if there is one, since we'll be incorporating
+	// it into the regex.
+	const baseFilename = filename.replace("./", "");
+	let regex = cachedRegex.get(baseFilename);
+	if (regex) {
+		return regex;
+	} else {
+		regex = new RegExp(`^(./)?${baseFilename}$`, "i");
+		cachedRegex.set(baseFilename, regex);
+		return regex;
+	}
+};
```

Comment:
[Refactor] I've seen this "make a Map for a single cache" pattern a bunch, I made a [`cached-factory`](https://www.npmjs.com/package/cached-factory) package for it. Up for using it to reduce code here? IMO this is a bit more readable:

```suggestion
import { CachedFactory } from "cached-factory";

const regexFactory = new CachedFactory(
	(baseFilename: string) => new RegExp(`^(./)?${baseFilename}$`, "i"),
);

const getCachedLocalFileRegex = (filename: string) =>
	regexFactory.get(filename.replace("./", ""));
```

(not a blocker at all)

Reply:
Right on.  Do you think it's worth the additional dependency?

Reply:
Updated to use `cached-factory`

Reply:
@JoshuaKGoldberg looks like `cached-factory` is ESM only, which is causing the cjs build to fail.  I'm going to go ahead and revert that change

Reply:
Ah, right, I forgot about that. Sounds like a wontfix comment to me. 🙂 thanks for the investigation!